### DISCUSSION TITLE
Is there a scenario where the reconciliation of a FluxCD Kustomization is abandoned partway through?

### DISCUSSION THREAD

**Discussion by cer** [fluxcd/flux2#4099] [Q&A]
I am using `flux: v2.0.1`

I have the following configuration:

* clusters/prod/sealedsecrets.yml - FluxCD Kustomization that installs seals secrets controller
* clusters/prod/apps.yml - FluxCD Kustomization pointing to apps/prod, depends on sealed secrets.
* apps/prod - includes the manifest for the SealedSecret for the repository credentials
* apps/base - has a HelmRelease and a HelmRepository that uses the unsealed secret
* clusters/dev/apps.yml - .. similar to above

I can successfully bootstrap both clusters on my fast local machine.
I have a (slower) GitHub Actions workflow that successfully bootstraps the dev cluster - the secret is unsealed, etc
But it then consistently fails to bootstrap the production cluster - the sealed secret does not exist

Is there a scenario where the reconciliation of a FluxCD Kustomization is abandoned partway through, e.g. the HelmRepository is never ready because the secret doesn't exist => reconciliation abandoned =>  which then causes the SealedSecret to not be created?

As an experiment I moved the SealedSecrets to a separate FluxCD Kustomization that `apps` depends upon. That appeared to work.

**Comment by stefanprodan** [comment]
Please see here how to organize your repo https://github.com/fluxcd/flux2-kustomize-helm-example

At minimum you would need 3 separate Flux Kustomizations with dependson relationship: controllers (CRDs), controller configs (CRs) and apps.

**Reply by stefanprodan** [reply]
By the way, what’s the reason for using SealedSecrets instead of the native Flux secrets management? Using Flux & Sops means one less controller to worry about.

**Reply by cer** [reply]
In my original problem description, I'd left out the sealed-secrets Flux Kustomization that installs Sealed Secrets Helm chart.
That was clearly necessary because Flux complained about the SealedSecret due to the missing CRD.

However, it's not obvious why the SealedSecret manifest needs to be in a different Kustomization from the HelmRepository that need the secret.

**Reply by stefanprodan** [reply]
Because Flux performs a dry run to validate the objects and since the definition is not in etcd the validation of the custom resources fails.

**Reply by cer** [reply]
Are you saying that the dry run might fail because the Secret referenced by the HelmRepository does not yet exist?

Is there a definition of "dry run to validate the objects" anywhere?

**Reply by stefanprodan** [reply]
I think I got your question wrong. If the SealedSecret CRD is present on the cluster, then it should work. Sealed secrets controller takes some some time to create the actual secret. Once that’s created the HelmRelease should succeed

**Comment by cer** [comment]
> I think I got your question wrong. If the SealedSecret CRD is present on the cluster, then it should work. Sealed secrets controller takes some some time to create the actual secret. Once that’s created the HelmRelease should succeed

That's what I've thought (and observed).
But for some reason, I have this situation:

* Dev - always works locally and in Github Actions
* Prod - always works locally but never in Github Actions - sealed secret is not created. And moving the sealed secret to a separate Flux Kustomization fixed the problem.
